plutus-benchmark/marlowe/test/Lib.hs

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE BlockArguments    #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Lib where
+
+import Control.Exception (throwIO)
+import Control.Monad.Except (runExceptT)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.IO qualified as Text.IO
+import Formatting qualified as F
+import PlutusCore.Default (DefaultFun, DefaultUni)
+import PlutusCore.Evaluation.Machine.ExMemory (CostingInteger)
+import PlutusCore.Test (runUPlcFull)
+import PlutusLedgerApi.V3 (ExBudget (..), ExCPU (..), ExMemory (..), fromSatInt)
+import System.Directory (createDirectoryIfMissing, removeFile)
+import System.FilePath qualified as Path
+import System.IO (Handle, IOMode (WriteMode), withFile)
+import Test.Tasty (TestName, TestTree)
+import Test.Tasty.Golden.Advanced (goldenTest2)
+import Text.Read (readMaybe)
+import Text.Tabular (Header (..), Properties (NoLine, SingleLine), Table (Table))
+import Text.Tabular.AsciiArt qualified as Tabular
+import UntypedPlutusCore (NamedDeBruijn)
+import UntypedPlutusCore.Core.Type qualified as UPLC
+import UntypedPlutusCore.Size qualified as UPLC
+
+-- | Measure the given program's execution budget and size.
+measureProgram
+  :: UPLC.Program NamedDeBruijn DefaultUni DefaultFun ()
+  -> IO (ExCPU, ExMemory, UPLC.Size)
+measureProgram program =
+  runExceptT (runUPlcFull [program]) >>= \case
+    Left err -> fail $ "Error evaluating UPLC program: " <> show err
+    Right (_term, ExBudget exCpu exMem, _logs) ->
+      pure (exCpu, exMem, UPLC.programSize program)
+
+{-| Compare the output file's contents against the golden file's contents
+after the given action has created the output file.
+-}
+goldenUplcMeasurements
+  :: TestName
+  -- ^ test name
+  -> FilePath
+  -- ^ path to the «golden» file (the file that contains correct output)
+  -> FilePath
+  -- ^ path to the output file
+  -> (Handle -> IO ())
+  -- ^ Given a file handle, action that writes measurements to it.
+  -> TestTree
+  {- ^ the test verifies that the output file contents is the same as
+  the golden file contents
+  -}
+goldenUplcMeasurements name goldenPath outputPath act =
+  goldenTest2
+    name
+    (Text.IO.readFile goldenPath)
+    (withFile outputPath WriteMode act >> Text.IO.readFile outputPath)
+    reportDifference
+    (createDirectoriesAndWriteFile goldenPath)
+    (removeFile outputPath)
+ where
+  reportDifference :: Text -> Text -> IO (Maybe String)
+  reportDifference expected actual = do
+    let parse = traverse parseLine . Text.lines
+    expectedResults <- parse expected
+    actualResults <- parse actual
+    if expectedResults == actualResults
+      then pure Nothing
+      else do
+        let (cpu0, mem0, size0) = aggregateResults expectedResults
+            (cpu1, mem1, size1) = aggregateResults actualResults
+
+        pure . Just $
+          "Actual execution budgets and sizes differ from expected ones:\n"
+            <> Tabular.render
```

Comment:
I generally just use `printf` for stuff like this.  It always makes me feel a bit guilty for some reason, but it has the merit of compactness.  I don't object to this though.